pkg/webhook/mutation_test.go

```diff
@@ -0,0 +1,220 @@
+package webhook
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ghodss/yaml"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	atypes "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func TestAssignMetaValidation(t *testing.T) {
+	tc := []struct {
+		Name          string
+		AssignMeta    string
+		ErrorExpected bool
+	}{
+		{
+			Name: "Valid Assign",
+			AssignMeta: `
+apiVersion: mutations.gatekeeper.sh
+kind: AssignMetadata
+metadata:
+  name: testAssignMeta
+spec:
+  location: metadata.labels.foo
+  parameters:
+    assign:
+      value: bar  
+`,
+			ErrorExpected: false,
+		},
+		{
+			Name: "Invalid Path",
+			AssignMeta: `
+apiVersion: mutations.gatekeeper.sh
+kind: AssignMetadata
+metadata:
+  name: testAssignMeta
+spec:
+  location: metadata.foo.bar
+  parameters:
+    assign:
+      value: bar  
+`,
+			ErrorExpected: true,
+		},
+		{
+			Name: "Invalid Assign",
+			AssignMeta: `
+apiVersion: mutations.gatekeeper.sh
+kind: AssignMetadata
+metadata:
+  name: testAssignMeta
+spec:
+  location: metadata.labels.bar
+  parameters:
+    assign:
+      foo: bar  
+`,
+			ErrorExpected: true,
+		},
+		{
+			Name: "Assign not a string",
+			AssignMeta: `
+apiVersion: mutations.gatekeeper.sh
+kind: AssignMetadata
+metadata:
+  name: testAssignMeta
+spec:
+  location: metadata.labels.bar
+  parameters:
+    assign:
+      value:
+        foo: bar  
+`,
+			ErrorExpected: true,
+		},
+		{
+			Name: "Assign no value",
+			AssignMeta: `
+apiVersion: mutations.gatekeeper.sh
+kind: AssignMetadata
+metadata:
+  name: testAssignMeta
+spec:
+  location: metadata.labels.bar
+  parameters:
+    assign:
+      zzz:
+        foo: bar  
+`,
+			ErrorExpected: true,
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.Name, func(t *testing.T) {
+			handler := mutationHandler{webhookHandler: webhookHandler{}}
+			b, err := yaml.YAMLToJSON([]byte(tt.AssignMeta))
+			if err != nil {
+				t.Fatalf("Error parsing yaml: %s", err)
+			}
+			review := atypes.Request{
+				AdmissionRequest: admissionv1beta1.AdmissionRequest{
+					Kind: metav1.GroupVersionKind{
+						Group:   "mutations.gatekeeper.sh",
+						Version: "v1alpha1",
+						Kind:    "AssignMetadata",
+					},
+					Object: runtime.RawExtension{
+						Raw: b,
+					},
+				},
+			}
+			_, err = handler.validateGatekeeperResources(context.Background(), review)
+			if err != nil && !tt.ErrorExpected {
+				t.Errorf("err = %s; want nil", err)
+			}
+			if err == nil && tt.ErrorExpected {
+				t.Error("err = nil; want non-nil")
+			}
+		})
+	}
+}
+
+func TestAssignValidation(t *testing.T) {
+	tc := []struct {
+		Name          string
+		Assign        string
+		ErrorExpected bool
+	}{
+		{
+			Name: "Valid Assign",
+			Assign: `
+apiVersion: mutations.gatekeeper.sh
+kind: Assign
+metadata:
+  name: goodAssign
+spec:
+  location: "spec.containers[name:test].foo"
```

Comment:
can we add in a test that makes sure we are denying `spec.containers[name: test].test` here?

If that test doesn't currently pass, we can comment it out and add a TODO to uncomment it, I just don't want to forget it

Reply:
Lol basically the example I put as a good example is not a good one :-)

Reply:
Added the validation and the test